src/libraries/System.Private.CoreLib/src/System/Text/Ascii.cs

```diff
@@ -0,0 +1,85 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace System.Text
+{
+    public static partial class Ascii
+    {
+        /// <summary>
+        /// Returns the index of the first non-ASCII byte in a buffer.
+        /// </summary>
+        /// <param name="buffer">The buffer to scan.</param>
+        /// <returns>The index in <paramref name="buffer"/> where the first non-ASCII
+        /// byte appears, or -1 if the buffer contains only ASCII bytes.</returns>
+        internal static unsafe int GetIndexOfFirstNonAsciiByte(ReadOnlySpan<byte> buffer)
+        {
+            if (buffer.IsEmpty)
+            {
+                return -1;
+            }
+
+            nuint bufferLength = (uint)buffer.Length;
+            fixed (byte* pBuffer = &MemoryMarshal.GetReference(buffer))
+            {
+                nuint idxOfFirstNonAsciiElement = GetIndexOfFirstNonAsciiByte(pBuffer, bufferLength);
+                Debug.Assert(idxOfFirstNonAsciiElement <= bufferLength);
+                return (idxOfFirstNonAsciiElement == bufferLength) ? -1 : (int)idxOfFirstNonAsciiElement;
+            }
+        }
+
+        /// <summary>
+        /// Returns the index of the first non-ASCII char in a buffer.
+        /// </summary>
+        /// <param name="buffer">The buffer to scan.</param>
+        /// <returns>The index in <paramref name="buffer"/> where the first non-ASCII
+        /// char appears, or -1 if the buffer contains only ASCII char.</returns>
+        internal static unsafe int GetIndexOfFirstNonAsciiChar(ReadOnlySpan<char> buffer)
+        {
+            if (buffer.IsEmpty)
+            {
+                return -1;
+            }
+
+            nuint bufferLength = (uint)buffer.Length;
+            fixed (char* pBuffer = &MemoryMarshal.GetReference(buffer))
+            {
+                nuint idxOfFirstNonAsciiElement = GetIndexOfFirstNonAsciiChar(pBuffer, bufferLength);
+                Debug.Assert(idxOfFirstNonAsciiElement <= bufferLength);
+                return (idxOfFirstNonAsciiElement == bufferLength) ? -1 : (int)idxOfFirstNonAsciiElement;
+            }
+        }
+
+        /// <summary>
+        /// Determines whether the provided value contains only ASCII bytes.
+        /// </summary>
+        /// <param name="value">The value to inspect.</param>
+        /// <returns>True if <paramref name="value"/> contains only ASCII bytes or is
+        /// empty; False otherwise.</returns>
+        public static unsafe bool IsValid(ReadOnlySpan<byte> value) => value.IsEmpty || GetIndexOfFirstNonAsciiByte(value) < 0;
```

Comment:
We expect empty to be so common that it's worth special-casing here and doing the check twice when it's not empty?  The very first thing GetIndexOfFirstNonAsciiByte does is the same IsEmpty check.